Fixing unblock condition for index create block

RS146BIJAY opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dbwiddis commented on server/src/main/java/org/opensearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -380,9 +380,17 @@ public void onNewInfo(ClusterInfo info) {
             && nodesOverHighThreshold.size() == nodes.size()) {
+            logger.warn(
+                "Enabling index create block on cluster as all nodes are breaching high disk watermark."
+                    + "Number of nodes above high watermark: {}. Total numbers of node: {}",
```
Need a space at end of previous string or start of this one.

RS146BIJAY replied:
Addressed the comments.

dbwiddis commented on server/src/test/java/org/opensearch/cluster/routing/allocation/DiskThresholdMonitorTests.java:
```diff
@@ -722,6 +718,111 @@ protected void setIndexCreateBlock(ActionListener<Void> listener, boolean indexC
+                .build()
+        );
+        Metadata metadata = Metadata.builder()
+            .put(
```
would like some comments explaining what these 3 tests are doing

RS146BIJAY replied:
Added comments ([Background Context](https://github.com/opensearch-project/OpenSearch/pull/5852)). Essentially the test validates that IndexCreateBlock only gets removed when any of the node is no longer breaching high watermark. Essentially earlier issue was this block was removed even though all the nodes were breaching high watermark, as condition for block removal was IndexCreateBlock should be present on the node and auto release of this block should be enabled (which is true by default). So if an IndexCreateBlock got added on the cluster (due to all node breaching high watermark), in the subsequent evaluation of DiskThresholdMonitor (after 15 sec), the earlier unblock condition will be true (even though all nodes are breaching high watermark) and IndexCreateBlock will be removed.

Let me know if this clarifies the isssue.

dbwiddis reviewed (commented):
Generally LGTM but I'm not sure I understand what's being fixed. The linked issue in the comments hasn't been replaced. The change log refers to a blank issue number.

RS146BIJAY commented:
> Generally LGTM but I'm not sure I understand what's being fixed. The linked issue in the comments hasn't been replaced. The change log refers to a blank issue number.

Addressed the comments. Essentially earlier issue was this block was removed even though all the nodes were breaching high watermark, as condition for block removal was IndexCreateBlock should be present on the node and auto release of this block should be enabled (which is true by default). So if an IndexCreateBlock got added on the cluster (due to all node breaching high watermark), in the subsequent evaluation of DiskThresholdMonitor (after 15 sec), the earlier unblock condition will be true (even though all nodes are breaching high watermark) and IndexCreateBlock will be removed.

Let me know if this clarifies the isssue.

BhumikaSaini-Amazon commented on CHANGELOG.md:
```diff
@@ -72,6 +72,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Fixed
 - Fix 'org.apache.hc.core5.http.ParseException: Invalid protocol version' under JDK 16+ ([#4827](https://github.com/opensearch-project/OpenSearch/pull/4827))
 - Fix compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
+- Fix condition to remove index create block ([#9437](https://github.com/opensearch-project/OpenSearch/pull/9437))
```
Will this not go in 2.x?

RS146BIJAY replied:
Yes. It will go. Addressed the comments

BhumikaSaini-Amazon commented on server/src/main/java/org/opensearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -380,9 +380,17 @@ public void onNewInfo(ClusterInfo info) {
         } else if (state.getBlocks().hasGlobalBlockWithId(Metadata.CLUSTER_CREATE_INDEX_BLOCK.id())
-            && diskThresholdSettings.isCreateIndexBlockAutoReleaseEnabled()) {
+            && diskThresholdSettings.isCreateIndexBlockAutoReleaseEnabled()
+            && nodesOverHighThreshold.size() != nodes.size()) {
```
Is considering a B/G relevant here? Does that change how this block guard should be interpreted? 

What I am trying to evaluate is if `nodesOverHighThreshold.size() < nodes.size()` is a better check here instead of `nodesOverHighThreshold.size() != nodes.size()` OR if they would end up being equivalent for our use case.

RS146BIJAY replied:
@BhumikaSaini-Amazon I agree `nodesOverHighThreshold.size() < nodes.size()` will be better check here.
@psychbot It does not matter, Right? In that case, we will check if active total number of nodes above high watermark is less than active total number of nodes. Once the affected nodes starts responding, `DiskThresholdMonitor` will start considering those nodes as well for evaluation and will add or remove nodes.

RS146BIJAY replied:
@BhumikaSaini-Amazon I dont think it is possible. Node will always be accounted in `nodesOverHighThreshold` if it is accounted in `nodes` and node is breaching high watermark. Since we are [iterating over all the nodes](https://github.com/opensearch-project/OpenSearch/blob/cf02b9677fde73056e071c6aaae2dd67b227ffe3/server/src/main/java/org/opensearch/cluster/routing/allocation/DiskThresholdMonitor.java#L216) and evaluating if it is breaching any watermarks.

BhumikaSaini-Amazon reviewed (approved):
LGTM, thanks for fixing this! Left a couple of questions

sachinpkale commented:
Overall changes look good. Please link bug issue in the PR description also resolve the conflict with `CHANGELOG.md`

RS146BIJAY commented on server/src/main/java/org/opensearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -380,9 +380,17 @@ public void onNewInfo(ClusterInfo info) {
             && nodesOverHighThreshold.size() == nodes.size()) {
+            logger.warn(
+                "Enabling index create block on cluster as all nodes are breaching high disk watermark. "
+                    + "Number of nodes above high watermark: {}. Total numbers of node: {}.",
```
Addressed the comments
